meta_review/handler.py

```diff
@@ -0,0 +1,440 @@
+
+        # save participants in memory
+        self.participants = dict()
+        for key, comment in self.comments.items():
+            # get author of each comment
+            author = comment['author']
+
+            # skip if user does not exist
+            # this happens when account is deleted from GitHub
+            if author['login']:
+                self.participants[author['login']] = author
+
+            if not comment.get('reactions'):
+                continue
+            for reaction in comment['reactions']:
+                # get user of each reaction
+                user = reaction['user']
+
+                # skip if user does not exist
+                if author['login']:
+                    self.participants[user['login']] = user
+
+    def handle(self):
+        """
+        Scrape data, process and store in database
+        """
+        load_all(self.participants, self.comments, self.reactions)
+
+        self.__check_comment_update()
+        self.__update_score()
+        self.__update_weight_factors()
+
+        # dump first to make use of built-in sort method
+        dump_all(self.participants, self.comments, self.reactions)
+        self.__update_rankings()
+
+        # dump first to make use of reverse query
+        dump_all(self.participants, self.comments, self.reactions)
+        self.__update_time()
+
+        dump_all(self.participants, self.comments, self.reactions)
+
+        self.logger.info('Meta Review System finishes.')
+
+    def __check_comment_update(self):
+        """
+        If reviewer updates their review comment after it has
+        been meta-reviewed, they will be slightly punished.
+        """
+        for key, comment in self.comments.items():
+            last_edited_at = comment.last_edited_at
+            author = comment.author
+
+            # skip if author does not exist
+            if not author:
+                continue
+
+            # if already punished during last iteration, don't punish again
+            # `last_active_at` attribute records time of last iteration
+            if (author.last_active_at and last_edited_at
+                    and author.last_active_at >= last_edited_at):
+                continue
+
+            reactions = comment.reaction_set.all()
+
+            need_punishment = False
+
+            # get reactions createdTime
+            for reaction in reactions:
+                if last_edited_at and last_edited_at > reaction.created_at:
+                    need_punishment = True
+
+            if need_punishment:
+                self.logger.info('%s updates review comment after it has been'
+                                 'meta-reviewed. 0.5 point deducted for'
+                                 'punishment. comment id: %s'
+                                 % (author.login, comment.id))
+                author.punishment += 0.5
+                author.score -= 0.5
+
+    def __update_time(self):
```

Comment:
why `__` prefix?

Reply:
This should be a private method IMO.

Reply:
private from whom?

It is an entirely open system, using public data.

What is the benefit of using `__` instead of weak private methods?
what is the risk model?

Anyone wanting to run this in a non-static system isnt going to have a more secure/stable system because of this, as some other part of code could access `.comments`, `.participants`, etc

OTOH, using `__` makes it harder to develop and test due to the mangling. Note that even the mangling doesnt stop access - just makes it a little bit harder.